### PR TITLE
Replace input handling in InputManager with DeviceSourceManager

### PR DESCRIPTION
**Pull request by PolygonalSun**
This PR contains several changes
- All observables have been removed from the WebDeviceInputSystem to improve parity with NativeDeviceInputSystem.
- DeviceSourceManager is now the primary source of input handling for the Input Manager.
- The engine now keeps track of an implementation of the DeviceSourceManager, instead of the DeviceInputSystem.
- The DeviceSourceManager class acts as a sort of wrapper for the engine-only singleton of InternalDeviceSourceManager. (Implemented this way because of backwards-compatibility)
- On web, touch no longer creates and deletes touch devices on a touch event.  The new system now creates slots for the maximum number of touches (usually 10, if touchscreen exists) and maintains them as inputs come in.  Anything that exceeds the max number of touches is ignored.

### CODE REVIEW

File: src/Engines/engine.ts [hunk 1]
```diff
@@ -391,9 +391,9 @@ export class Engine extends ThinEngine {
     public isPointerLock = false;
 
     /**
-     * Stores instance of DeviceInputSystem
+     * @hidden
      */
-    public deviceInputSystem: DeviceInputSystem;
+    public _deviceSourceManager: InternalDeviceSourceManager;
```
Comment (bghgary) [inline src/Engines/engine.ts:396]: This seems backwards to me. Why not augment the engine from the DeviceSourceManager side and then you won't have to expose InternalDeviceSourceManager at all. You can use engine.onDisposeObservable to dispose the instance.

File: src/DeviceInput/InputDevices/internalDeviceSourceManager.ts [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+import { Observable, Observer } from '../../Misc/observable';
+import { IDeviceEvent, IDeviceInputSystem } from './inputInterfaces';
+import { NativeDeviceInputSystem } from './nativeDeviceInputSystem';
+import { WebDeviceInputSystem } from './webDeviceInputSystem';
+import { DeviceSource } from './deviceSource';
+import { INative } from '../../Engines/Native/nativeInterfaces';
+
+declare const _native: INative;
+
+/** @hidden */
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/internalDeviceSourceManager.ts:14]: I believe once you put `@hidden` the top level class, you don't need to add `@hidden` for each child public member, but this won't matter if we can do what I said in the other comment since this class will then be private.

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts [hunk 3]
```diff
@@ -89,7 +89,7 @@ export class WebDeviceInputSystemImpl implements IDeviceInputSystem {
     /**
      * Configures events to work with an engine's active element
      */
-    public configureEvents() {
+    private configureEvents() {
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/webDeviceInputSystem.ts:92]:
```suggestion
    private _configureEvents() {
```

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts [hunk 4]
```diff
@@ -141,7 +141,7 @@ export class WebDeviceInputSystemImpl implements IDeviceInputSystem {
      * @param deviceType Type of device to check for
      * @returns bool with status of device's existence
      */
-    public isDeviceAvailable(deviceType: DeviceType) {
+    public isDeviceAvailable(deviceType: DeviceType): boolean {
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/webDeviceInputSystem.ts:144]: nit: I would recommend adding return types to all of the functions not just this one.

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts [hunk 5]
```diff
@@ -330,6 +330,15 @@ export class WebDeviceInputSystemImpl implements IDeviceInputSystem {
      * Handle all actions that come from pointer interaction
      */
     private _handlePointerActions() {
+        this._maxTouchPoints = navigator.maxTouchPoints || 0;
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/webDeviceInputSystem.ts:333]: Is it safe to use navigator without checking that it is available? Maybe to be safe just add a check `DomManagement.IsNavigatorAvailable()`.

```suggestion
        this._maxTouchPoints = (DomManagement.IsNavigatorAvailable() && navigator.maxTouchPoints) || 0;
```

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts [hunk 6]
```diff
@@ -60,8 +60,8 @@ export class WebDeviceInputSystemImpl implements IDeviceInputSystem {
     private _isUsingFirefox = navigator && navigator.userAgent && navigator.userAgent.indexOf("Firefox") !== -1;
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/webDeviceInputSystem.ts:60]: Not sure this is the right way to check for navigator existence.

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts [hunk 7]
```diff
@@ -406,8 +415,16 @@ export class WebDeviceInputSystemImpl implements IDeviceInputSystem {
             let deviceSlot = (deviceType === DeviceType.Mouse) ? 0 : evt.pointerId;
 
             if (deviceType === DeviceType.Touch) {
-                deviceSlot = this._rollingTouchId++;
-                this._activeTouchIds[deviceSlot] = evt.pointerId;
+                let idx = this._activeTouchIds.indexOf(-1);
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/webDeviceInputSystem.ts:418]: nit:
```suggestion
                const idx = this._activeTouchIds.indexOf(-1);
```

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts [hunk 8]
```diff
@@ -406,8 +415,16 @@ export class WebDeviceInputSystemImpl implements IDeviceInputSystem {
...
-                deviceSlot = this._rollingTouchId++;
-                this._activeTouchIds[deviceSlot] = evt.pointerId;
+                let idx = this._activeTouchIds.indexOf(-1);
+
+                if (idx >= 0) {
+                    deviceSlot = idx;
+                    this._activeTouchIds[idx] = evt.pointerId;
+                }
+                else {
+                    // We can't find an open slot to store new pointer so just return (can only support max number of touches)
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/webDeviceInputSystem.ts:425]: Maybe log a warning?

**Review by sebavan — COMMENTED** [review]
@bghgary said it all :-)

File: src/Engines/engine.ts (hunk 1 above)
Comment (PolygonalSun) [reply src/Engines/engine.ts:396]: Discussed offline

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts (hunk 8 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:425]: Added warning

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts (hunk 7 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:418]: changed

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts (hunk 5 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:333]: changed

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts (hunk 4 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:144]: added

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts (hunk 3 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:92]: changed

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts (hunk 6 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:60]: changed

File: src/DeviceInput/InputDevices/internalDeviceSourceManager.ts (hunk 2 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/internalDeviceSourceManager.ts:14]: Talked offline

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts [hunk 9]
```diff
@@ -636,8 +636,6 @@ export class WebDeviceInputSystem implements IDeviceInputSystem {
                         deviceEvent.previousState = 1;
                         this.onInputChanged(deviceEvent);
 
-                        this._unregisterDevice(DeviceType.Touch, deviceSlot);
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/webDeviceInputSystem.ts:620]: why was this removed?

Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:620]: So I had originally removed this because I had a rolling system of active ids and wanted to prevent the any unnecessary array deletions for touch.  After doing a couple of edge tests, I've found that I need to have something still trigger the onDeviceDisconnected functionality so I'm going to add a direct call to it in place of this.

**Comment by sebavan** [comment]
I ll let @bghgary and @deltakosh check the final bits (tech details) this week but was happy with the overall approach on our Friday chat.

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts [hunk 10]
```diff
@@ -41,7 +45,7 @@ export class WebDeviceInputSystemImpl implements IDeviceInputSystem {
     private _pointerActive: boolean = false;
     private _elementToAttachTo: HTMLElement;
     private _engine: Engine;
```
Comment (bghgary) [inline src/DeviceInput/InputDevices/webDeviceInputSystem.ts:47]: nit: could be readonly. maybe can do a pass on this later.

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts (hunk 9 above)
Comment (bghgary) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:620]: Are we confident the array won't grow too big?

**Review by bghgary — COMMENTED** [review]
Looks good. Just one last verification.

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts (hunk 9 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:620]: The touch array should never grow any larger than the value of _maxTouchPoints which should be no greater than 10 (platform-dependent).

File: src/DeviceInput/InputDevices/webDeviceInputSystem.ts (hunk 10 above)
Comment (PolygonalSun) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts:47]: Easy enough to change

Comment (bghgary) [reply src/DeviceInput/InputDevices/webDeviceInputSystem.ts]: There are more that can be readonly, but no need to do this now.

**Review by bghgary — APPROVED** [review]
